### PR TITLE
Move PR validation logic from the official build to a separate yml

### PR DESCRIPTION
**Pull request by JoeRobich**
- Removes the ability to run the Official build on DevDiv
- Removes the Set Build Number step since we can use the given build number
- Removes the PR validation into its own yml to run in its own pipeline
- Requires the PR's head SHA to be provided and match when running PR Validation. 

cc: @genlu, @RikkiGibson 

Test PR validation with too short SHA: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4865094&view=results
Test PR validation with invalid SHA: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4865097&view=results
Test PR validation with correct SHA: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4865173&view=results
Test official build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1184266&view=results

### CODE REVIEW

File: azure-pipelines-official.yml [hunk 1]
```diff
@@ -54,31 +47,13 @@ stages:
   - job: OfficialBuild
     displayName: Official Build
     timeoutInMinutes: 360
-    # Conditionally set build pool so we can share this YAML when building with different pipeline (devdiv vs dnceng)
+    # Conditionally set build pool so we can share this YAML when building with different pipeline
```
Comment (genlu) [inline azure-pipelines-official.yml:50]: We probably can remove all the checks for dnceng pipeline from this file too, or if we want to be sure, have a single, separate task at the beginning that fails the build if not running in dnceng.

**Comment by genlu** [comment]
>       displayName: Merge main-vs-deps into source branch

not required for this PR, but we might want to move main validation to the new validation pipeline as well

---
Refers to: azure-pipelines-official.yml:77 in 58e91fc. [](commit_id = 58e91fc0959ed750e55dd410bdb3f08dd9e709e9, deletion_comment = False)

**Comment by JoeRobich** [comment]
> not required for this PR, but we might want to move main validation to the new validation pipeline as well

@genlu I like the suggestion of moving main validation out of the official build. I wouldn't want to add it to the pr validation yml, so maybe there are multiple which can share a common build template.

File: azure-pipelines-pr-validation.yml [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                   -officialBuildId $(Build.BuildNumber)
+                   -officialSkipTests $(SkipTests)
+                   -officialSkipApplyOptimizationData $(SkipApplyOptimizationData)
+                   -officialSourceBranchName $(SourceBranchName)
+                   -officialIbcDrop $(IbcDrop)
+                   -officialVisualStudioDropAccessToken $(_DevDivDropAccessToken)
+                   /p:RepositoryName=$(Build.Repository.Name)
+                   /p:VisualStudioDropName=$(VisualStudio.DropName)
+                   /p:DotNetSignType=$(SignType)
+                   /p:DotNetPublishToBlobFeed=false
```
Comment (RikkiGibson) [inline azure-pipelines-pr-validation.yml:117]: Should we delete the unused arguments whose names start with `DotNetPublishBlobFeed`, `DotNetSymbolServerToken`, etc?

File: azure-pipelines-pr-validation.yml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      displayName: Setting SourceBranchName variable
+      condition: succeeded()
+
+    - task: tagBuildOrRelease@0
+      displayName: Tag PR validation build
+      inputs:
+        type: 'Build'
+        tags: |
+          PRValidationBuild
+          PRNumber:$(PRNumber)
```
Comment (RikkiGibson) [inline azure-pipelines-pr-validation.yml:58]: Should we also include a `CommitSHA:$(CommitSHA)` tag, for example, since it appears that the PR commit SHA is not related to the commit ID associated with the build.

File: eng/setup-pr-validation.ps1 [hunk 4]
```diff
@@ -1,24 +1,40 @@
...
+      Write-Host "##vso[task.LogIssue type=error;]The PR Commit SHA must be at least 7 characters long."
+      exit 1
+    }
+
+    Write-Host "Getting the hash of refs/pull/$prNumber/head..."
+    $remoteRef = git ls-remote origin refs/pull/$prNumber/head
+    Write-Host ($remoteRef | Out-String)
+
+    $prHeadSHA = $remoteRef.Split()[0]
+    if (!$prHeadSHA.StartsWith($commitSHA)) {
```
Comment (RikkiGibson) [inline eng/setup-pr-validation.ps1:27]: I think it might be reasonable to simply check if the `$commitSHA` is an ancestor of the `$prHeadSHA` because there may be a scenario where the user wants to validate one of the earlier commits in their PR.

File: eng/setup-pr-validation.ps1 [hunk 5]
```diff
@@ -1,24 +1,40 @@
...
+      Write-Host "##vso[task.LogIssue type=error;]The PR's Head SHA ($prHeadSHA) does not begin with the specified commit SHA ($commitSHA). Unreviewed changes may have been pushed to the PR."
+      exit 1
+    }
+
     Write-Host "Setting up the build for PR validation by merging refs/pull/$prNumber/merge into $sourceBranchName..."
     git pull origin refs/pull/$prNumber/merge
```
Comment (RikkiGibson) [inline eng/setup-pr-validation.ps1:33]: This seems like a race condition. The remote SHA may have changed in between `ls-remote` and `pull`. Consider pulling the `$commitSHA` instead.

**Comment by RikkiGibson** [comment]
> Test official build: https://dev.azure.com/dnceng/internal/_build/results?buildId=1184266&view=results

It looks like this build failed. It's not clear to me if it could be related to the official yml changes.

File: eng/setup-pr-validation.ps1 (hunk 4 above)
Comment (JoeRobich) [reply eng/setup-pr-validation.ps1:27]: Yeah, I thought over this, but wasn't sure what the scenario would be. We obviously couldn't take half a PR. Also, It isn't the same as pulling pull/##/merge, since it wouldn't contain changes from the target branch.

Happy to defer this as a future enhancement though, once we work out how it should work.

File: eng/setup-pr-validation.ps1 (hunk 5 above)
Comment (JoeRobich) [reply eng/setup-pr-validation.ps1:33]: I mean technically true, but would be very tight. 

Happy to defer this as a future enhancement where we also include merging the target branch in with a specific commit sha from the PR branch.

**Comment by JoeRobich** [comment]
> It looks like this build failed. It's not clear to me if it could be related to the official yml changes.

Failed due to missing optprof. Will rerun specifying a "potentially" good optprof drop https://dev.azure.com/dnceng/internal/_build/results?buildId=1188427&view=results

```
D:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21230.2\tools\OptimizationData.targets(110,5): error : No optimization data found for assemblies: D:\workspace\_work\1\s\artifacts\obj\Microsoft.VisualStudio.LanguageServices.CodeLens\Release\net472\Microsoft.VisualStudio.LanguageServices.CodeLens.dll [D:\workspace\_work\1\s\src\VisualStudio\CodeLens\Microsoft.VisualStudio.LanguageServices.CodeLens.csproj]
```

File: azure-pipelines-pr-validation.yml (hunk 3 above)
Comment (JoeRobich) [reply azure-pipelines-pr-validation.yml:58]: I like this suggestion

File: azure-pipelines-pr-validation.yml (hunk 2 above)
Comment (JoeRobich) [reply azure-pipelines-pr-validation.yml:117]: Maybe?  I'll give it a shot. =p

File: azure-pipelines-official.yml (hunk 1 above)
Comment (JoeRobich) [reply azure-pipelines-official.yml]: Removed!

File: eng/setup-pr-validation.ps1 (hunk 4 above)
Comment (RikkiGibson) [reply eng/setup-pr-validation.ps1:27]: The scenario is that my PR has an RPS regression and I suspect a specific commit in the PR introduced the regression. So I may want to validate the specific commit as well as the one before it. That said, there is no need to introduce this capability, since it would be new.

File: eng/setup-pr-validation.ps1 (hunk 5 above)
Comment (RikkiGibson) [reply eng/setup-pr-validation.ps1:33]: It feels like the risk of this may be less if we fetch all the refs we want in one invocation.

**Review by RikkiGibson — APPROVED** [review]

File: eng/setup-pr-validation.ps1 [hunk 6]
```diff
@@ -1,24 +1,40 @@
...
     }
+
+    Write-Host "Validating the PR head matches the specified commit SHA ($commitSHA)..."
+    if ($commitSHA.Length -lt 7) {
+      Write-Host "##vso[task.LogIssue type=error;]The PR Commit SHA must be at least 7 characters long."
+      exit 1
+    }
+
+    Write-Host "Getting the hash of refs/pull/$prNumber/head..."
+    $remoteRef = git ls-remote origin refs/pull/$prNumber/head
```
Comment (genlu) [inline eng/setup-pr-validation.ps1:23]: >refs/pull/$prNumber/head [](start = 38, length = 24)

Would it be an issue that we check for the SHA of refs/pull/$prNumber/head but later do a separate pull from another branch refs/pull/$prNumber/merge

Comment (RikkiGibson) [reply eng/setup-pr-validation.ps1:23]: Is this the same concern as https://github.com/dotnet/roslyn/pull/54053#discussion_r652046807?

**Review by genlu — APPROVED** [review]
:shipit:

File: eng/setup-pr-validation.ps1 (hunk 6 above)
Comment (JoeRobich) [reply eng/setup-pr-validation.ps1:23]: We've been using the merge branch since it will bring along changes from the PR's target branch and give us the most accurate representation of the PR merging and flowing into the release branch. The commit sha is intended to ensure that the PR branch is still in the same state as it was when reviewed by the user starting the validation.
